utils/replicaset/canary.go

```diff
@@ -160,23 +160,23 @@ func CalculateReplicaCountsForCanary(rollout *v1alpha1.Rollout, newRS *appsv1.Re
 		}
 	}
 
-	minAvailableReplicaCount := rolloutSpecReplica - MaxUnavailable(rollout)
+	if GetReplicaCountForReplicaSets(oldRSs) > 0 {
+		// If any older ReplicaSets exist, we should scale those down first, before even considering
+		// scaling down the newRS or stableRS
+		return newRSReplicaCount, stableRSReplicaCount
+	}
 
-	totalAvailableOlderReplicaCount := GetAvailableReplicaCountForReplicaSets(oldRSs)
-	scaleDownCount := GetReplicasForScaleDown(newRS) + GetReplicasForScaleDown(stableRS) + totalAvailableOlderReplicaCount - minAvailableReplicaCount
+	minAvailableReplicaCount := rolloutSpecReplica - MaxUnavailable(rollout)
+	replicasToScaleDown := GetReplicasForScaleDown(newRS, false) + GetReplicasForScaleDown(stableRS, true)
 
-	if scaleDownCount <= 0 {
+	if replicasToScaleDown <= minAvailableReplicaCount {
 		// Cannot scale down stableRS or newRS without going below min available replica count
 		return newRSReplicaCount, stableRSReplicaCount
 	}
 
-	if scaleDownCount <= totalAvailableOlderReplicaCount {
-		//Need to scale down older replicas before scaling down the newRS or stableRS.
-		return newRSReplicaCount, stableRSReplicaCount
-	}
-	scaleDownCount = scaleDownCount - totalAvailableOlderReplicaCount
+	scaleDownCount := replicasToScaleDown - minAvailableReplicaCount
 
-	if newRS != nil && *newRS.Spec.Replicas > desiredNewRSReplicaCount && scaleDownCount > 0 {
+	if newRS != nil && *newRS.Spec.Replicas > desiredNewRSReplicaCount {
```

Comment:
`&& scaleDownCount > 0` check is/was unnecessary because `replicasToScaleDown <= minAvailableReplicaCount` covers that.